File: src/main/java/rocks/sakira/gentlefawn/register/Blocks.java
package rocks.sakira.gentlefawn.register;

import net.minecraft.block.Block;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import rocks.sakira.gentlefawn.GentleFawn;
import rocks.sakira.gentlefawn.block.DeerSkullBlock;
import rocks.sakira.gentlefawn.block.DeerWallSkullBlock;

import static net.minecraft.block.Blocks.DRAGON_HEAD;

public class Blocks {
    public static final DeferredRegister<Block> REGISTER = DeferredRegister.create(ForgeRegistries.BLOCKS, GentleFawn.MOD_ID);

    public static final RegistryObject<Block> DOE_HEAD = REGISTER.register(
            "doe_head",

            () -> new DeerSkullBlock(
                    DeerSkullBlock.Types.DOE,
                    Block.Properties.from(DRAGON_HEAD)
            )
    );

    public static final RegistryObject<Block> DOE_WALL_HEAD = REGISTER.register(
            "doe_wall_head",

            () -> new DeerWallSkullBlock(
                    DeerSkullBlock.Types.DOE,
                    Block.Properties.from(DRAGON_HEAD)
                        .lootFrom(DOE_HEAD.get())
            )
    );

    public static final RegistryObject<Block> BUCK_HEAD = REGISTER.register(
            "buck_head",

            () -> new DeerSkullBlock(
                    DeerSkullBlock.Types.BUCK,
                    Block.Properties.from(DRAGON_HEAD)
            )
    );

    public static final RegistryObject<Block> BUCK_WALL_HEAD = REGISTER.register(
            "buck_wall_head",

            () -> new DeerWallSkullBlock(
                    DeerSkullBlock.Types.BUCK,
                    Block.Properties.from(DRAGON_HEAD)
                            .lootFrom(BUCK_HEAD.get())
            )
    );
}

File: src/main/kotlin/io/github/faecraft/gentlefawn/register/Blocks.kt
package io.github.faecraft.gentlefawn.register

import io.github.faecraft.gentlefawn.Common.MOD_ID
import io.github.faecraft.gentlefawn.block.UtilSkullBlock
import io.github.faecraft.gentlefawn.block.UtilSkullWallBlock
import io.github.faecraft.gentlefawn.enum.DeerSkullType
import net.fabricmc.fabric.api.`object`.builder.v1.block.FabricBlockSettings
import net.minecraft.block.Block
import net.minecraft.block.Blocks
import net.minecraft.block.SkullBlock
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

object Blocks {
    val DOE_HEAD = createSkullBlock(DeerSkullType.DOE)
    val DOE_HEAD_WALL = createSkullWallBlock(DeerSkullType.DOE, DOE_HEAD)

    val BUCK_HEAD = createSkullBlock(DeerSkullType.BUCK)
    val BUCK_HEAD_WALL = createSkullWallBlock(DeerSkullType.BUCK, BUCK_HEAD)

    fun register() {
        register(DOE_HEAD, "doe_head")
        register(DOE_HEAD_WALL, "doe_wall_head")

        register(BUCK_HEAD, "buck_head")
        register(BUCK_HEAD_WALL, "buck_wall_head")
    }

    private fun register(block: Block, id: String) {
        Registry.register(Registry.BLOCK, Identifier(MOD_ID, id), block)
    }

    private fun createSkullBlock(type: SkullBlock.SkullType) = UtilSkullBlock(
        type,

        FabricBlockSettings
            .copyOf(Blocks.DRAGON_HEAD)
    )

    private fun createSkullWallBlock(type: SkullBlock.SkullType, dropsLike: Block) = UtilSkullWallBlock(
        type,

        FabricBlockSettings
            .copyOf(Blocks.DRAGON_HEAD)
            .dropsLike(dropsLike)
    )
}
